src/code/z_camera.c

```diff
@@ -7858,140 +7872,181 @@ void Camera_Finish(Camera* camera) {
         return -1;
     }
 
-    if (!((sCameraSettings[camera->setting].unk_00 & 0x3FFFFFFF) & (1 << mode))) {
-        if (mode == CAM_MODE_FIRST_PERSON) {
+    if (!((sCameraSettings[camera->setting].unk_00 & 0x3FFFFFFF) & (1 << requestedMode))) {
+        if (requestedMode == CAM_MODE_FIRST_PERSON) {
             osSyncPrintf("camera: error sound\n");
             func_80078884(NA_SE_SY_ERROR);
         }
 
         if (camera->mode != CAM_MODE_NORMAL) {
             osSyncPrintf(VT_COL(YELLOW, BLACK) "camera: change camera mode: force NORMAL: %s %s refused\n" VT_RST,
-                         sCameraSettingNames[camera->setting], sCameraModeNames[mode]);
+                         sCameraSettingNames[camera->setting], sCameraModeNames[requestedMode]);
             camera->mode = CAM_MODE_NORMAL;
             Camera_CopyDataToRegs(camera, camera->mode);
-            func_8005A02C(camera);
-            return 0xC0000000 | mode;
-        } else {
-            camera->behaviorFlags |= CAM_BEHAVIOR_MODE_2;
-            camera->behaviorFlags |= CAM_BEHAVIOR_MODE_1;
-            return 0;
-        }
-    } else {
-        if (mode == camera->mode && flags == 0) {
-            camera->behaviorFlags |= CAM_BEHAVIOR_MODE_2;
-            camera->behaviorFlags |= CAM_BEHAVIOR_MODE_1;
-            return -1;
-        }
-        camera->behaviorFlags |= CAM_BEHAVIOR_MODE_2;
-        camera->behaviorFlags |= CAM_BEHAVIOR_MODE_1;
-        Camera_CopyDataToRegs(camera, mode);
-        modeChangeFlags = 0;
-        switch (mode) {
-            case CAM_MODE_FIRST_PERSON:
-                modeChangeFlags = 0x20;
-                break;
-            case CAM_MODE_Z_TARGET_UNFRIENDLY:
-                modeChangeFlags = 4;
-                break;
-            case CAM_MODE_Z_TARGET_FRIENDLY:
-                if (camera->target != NULL && camera->target->id != ACTOR_EN_BOOM) {
-                    modeChangeFlags = 8;
-                }
-                break;
-            case CAM_MODE_Z_PARALLEL:
-            case CAM_MODE_TALK:
-            case CAM_MODE_Z_AIM:
-            case CAM_MODE_Z_LEDGE_HANG:
-            case CAM_MODE_PUSH_PULL:
-                modeChangeFlags = 2;
-                break;
+            Camera_SetNewModeStateFlags(camera);
+            return 0xC0000000 | requestedMode;
         }
 
-        switch (camera->mode) {
-            case CAM_MODE_FIRST_PERSON:
-                if (modeChangeFlags & 0x20) {
-                    camera->animState = 10;
-                }
-                break;
-            case CAM_MODE_Z_PARALLEL:
-                if (modeChangeFlags & 0x10) {
-                    camera->animState = 10;
-                }
-                modeChangeFlags |= 1;
-                break;
-            case CAM_MODE_CHARGE:
-                modeChangeFlags |= 1;
+        camera->behaviorFlags |= CAM_BEHAVIOR_MODE_VALID;
+        camera->behaviorFlags |= CAM_BEHAVIOR_MODE_SUCCESS;
+        return CAM_MODE_NORMAL;
```

Comment:
This function does return the `mode` in its lower bits, and different "flags" in its upper bits. For example,
`return requestedMode | 0x80000000;`

Now, this block containing:
`return CAM_MODE_NORMAL;`
is only reached if `camera->mode == CAM_MODE_NORMAL;` already, because otherwise it's caught by the `if (camera->mode != CAM_MODE_NORMAL) {`

Reply:
I'm kinda stuck on this but idk I think this is just overall jank/bad code

this spot is the only return that's >= 0, which is afaik only checked here `if (Camera_RequestModeImpl(camera, camera->mode, true) >= 0) {`

and checks on `func_8083AD4C` are just == or != 0, whereas if it checked the mode I'd expect `func_8083AD4C() & low_bits_mask`

Reply:
I see, because it's `s32` return, the `| 0x80000000` and `| 0xC0000000` make the return value negative, so for the block:
```c
if (Camera_RequestModeImpl(camera, camera->mode, true) >= 0) {
    Camera_CopyDataToRegs(camera, camera->mode);
}
```
in `Camera_RequestSettingImpl`, it has to be 0 via `return 0` as it's impossible to return a positive value?